catalog/app/utils/GraphQL/wrappers.ts

```diff
@@ -0,0 +1,145 @@
+import * as R from 'ramda'
+import * as React from 'react'
+import * as urql from 'urql'
+import * as Sentry from '@sentry/react'
+
+import log from 'utils/Logging'
+import { BaseError } from 'utils/error'
+import useMemoEq from 'utils/useMemoEq'
+
+export class OperationError extends BaseError {}
+
+export class QueryError<Data = any> extends OperationError {
+  constructor(result?: urql.UseQueryState<Data, any>) {
+    super('GraphQL query error')
+    this.result = result
+  }
+}
+
+export class MutationError<Data, Variables> extends OperationError {
+  constructor(result?: urql.OperationResult<Data, Variables>) {
+    super('GraphQL mutation error')
+    this.result = result
+  }
+}
+
+interface FoldOptions<Data, OnData, OnFecthing, OnError = never> {
+  data: (data: Data, result: urql.UseQueryState<Data, any>) => OnData
+  fetching: (result: urql.UseQueryState<Data, any>) => OnFecthing
+  error?: (
+    error: urql.CombinedError | QueryError<Data>,
+    result: urql.UseQueryState<Data, any>,
+  ) => OnError
+  partial?: boolean
+  silent?: boolean
+}
+
+export const foldC =
+  <Data, OnData, OnFetching, OnError = never>({
+    data,
+    error,
+    fetching,
+    partial = false,
+    silent = false,
+  }: FoldOptions<Data, OnData, OnFetching, OnError>) =>
+  (result: urql.UseQueryState<Data, any>) => {
+    const isPartial = result.operation?.context?.meta?.cacheOutcome === 'partial'
+    if (!partial && isPartial) return fetching(result)
+
+    if (result.data) {
+      if (!silent && result.error) {
+        log.warn('Non-critical error while executing the query:', result.error)
+        Sentry.captureException(result.error, { extra: { result } })
+      }
+      return data(result.data, result)
+    }
+
+    if (result.fetching) return fetching(result)
+
+    const err = result.error || new QueryError(result)
+    if (error) return error(err, result)
+    if (!silent) {
+      log.error('Critical error while executing the query:', err)
+      Sentry.captureException(err, { extra: { result } })
+    }
+    throw error
+  }
+
+export const fold = <Data, OnData, OnFetching, OnError = never>(
```

Comment:
I think that for core functions like these we can provide a return type as well, even typescript can guess it.
Because it will be some sort of documentation

Reply:
i can see the return types by using the "hover" functionality of the language server, it's deduced by TS automatically, and when using a language server or some doc generators, the type should be displayed just fine. tho i agree that it may be a good idea to explicitly add some return type annotations for documentation purposes